plugin/plugin.go

```diff
@@ -0,0 +1,160 @@
+	// SIGNATURE_GENERATOR capability.
+	CommandGenerateSignature Command = "generate-signature"
+
+	// CommandGenerateEnvelope is the name of the plugin command
+	// which must be supported by every plugin that has the
+	// SIGNATURE_ENVELOPE_GENERATOR capability.
+	CommandGenerateEnvelope Command = "generate-envelope"
+)
+
+// Capability returns the capability associated to the command.
+func (c Command) Capability() Capability {
+	switch c {
+	case CommandGenerateSignature:
+		return CapabilitySignatureGenerator
+	case CommandGenerateEnvelope:
+		return CapabilityEnvelopeGenerator
+	default:
+		return ""
+	}
+}
+
+// Capability returns the response associated to the command.
+func (c Command) NewResponse() interface{} {
+	switch c {
+	case CommandGetMetadata:
+		return new(Metadata)
+	case CommandGenerateSignature:
+		return new(GenerateSignatureResponse)
+	case CommandGenerateEnvelope:
+		return new(GenerateEnvelopeResponse)
+	default:
+		return nil
+	}
+}
+
+// Capability is a feature available in the plugin contract.
+type Capability string
+
+const (
+	// CapabilitySignatureGenerator is the name of the capability
+	// which should support a plugin to support generating signatures.
+	CapabilitySignatureGenerator Capability = "SIGNATURE_GENERATOR"
+
+	// CapabilityEnvelopeGenerator is the name of the capability
+	// which should support a plugin to support generating envelope signatures.
+	CapabilityEnvelopeGenerator Capability = "SIGNATURE_ENVELOPE_GENERATOR"
+)
+
+// GenerateSignatureRequest contains the parameters passed in a generate-signature request.
+// All parameters are required.
+type GenerateSignatureRequest struct {
+	ContractVersion string
+	KeyName         string
+	KeyID           string
+}
+
+func (req *GenerateSignatureRequest) Command() Command {
+	return CommandGenerateSignature
+}
+
+func (req *GenerateSignatureRequest) Validate() error {
+	if req == nil {
+		return errors.New("nil request")
+	}
+	if req.ContractVersion == "" {
+		return errors.New("empty contractVersion")
+	}
+	if req.KeyName == "" {
+		return errors.New("empty keyName")
+	}
+	if req.KeyID == "" {
+		return errors.New("empty keyId")
+	}
+	return nil
+}
+
+// GenerateSignatureResponse is the response of a generate-signature request.
+type GenerateSignatureResponse struct {
+	// The same key id as passed in the request.
+	KeyID string `json:"keyId"`
+
+	// Base64 encoded signature.
+	Signature string `json:"signature"`
+
+	// One of following supported signing algorithms:
+	// https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#algorithm-selection
+	SigningAlgorithm string `json:"signingAlgorithm"`
+
+	// Ordered list of certificates starting with leaf certificate
+	// and ending with root certificate.
+	CertificateChain []string `json:"certificateChain"`
+}
+
+// GenerateEnvelopeRequest contains the parameters passed in a generate-envelop request.
+// All parameters are required.
+type GenerateEnvelopeRequest struct {
```

Comment:
Missing the `payload` field defined in [spec](https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md#user-content-signature-envelope-generator).